octobot_trading/producers/simulator/ohlcv_updater_simulator.py

```diff
@@ -33,34 +33,59 @@ def __init__(self, channel, importer):
-                for pair in self.channel.exchange_manager.exchange_config.traded_symbol_pairs:
-                    # use timestamp - 1 for superior timestamp to avoid select of a future candle
+            for pair in self.channel.exchange_manager.exchange_config.traded_symbol_pairs:
+                for time_frame in self.channel.exchange_manager.exchange_config.traded_time_frames:
+                    # Use last_timestamp_pushed + 1 for inferior timestamp to avoid select of an already selected candle
                     # (selection is <= and >=)
+                    # Use timestamp + self.future_candle_sec_length to include the future candle on the future candles
+                    # time frame that will be sorted in exchange simulator for later uses.
                     ohlcv_data: list = await self.exchange_data_importer.get_ohlcv_from_timestamps(
                         exchange_name=self.exchange_name,
                         symbol=pair,
                         time_frame=time_frame,
-                        inferior_timestamp=self.last_timestamp_pushed,
-                        superior_timestamp=timestamp - 1)
+                        inferior_timestamp=self.last_timestamp_pushed + 1,
+                        superior_timestamp=timestamp + (self.future_candle_sec_length
+                                                        if self.future_candle_time_frame is time_frame else 0)
+                    )
                     if ohlcv_data:
-                        await self.push(time_frame, pair, [ohlcv[-1] for ohlcv in ohlcv_data], partial=True)
-
+                        current_candle_index = 0
+                        if self.future_candle_time_frame is time_frame:
+                            if ohlcv_data[0][-1][PriceIndexes.IND_PRICE_TIME.value] == timestamp:
+                                # register future candle
+                                self.channel.exchange.current_future_candles[time_frame.value] = ohlcv_data[0][-1]
+                                # do not push future candle
+                                current_candle_index = 1
+                            else:
+                                # if no future candle available
+                                # (end of backtesting of missing data: reset future candle)
+                                self.channel.exchange.current_future_candles[time_frame.value] = None
+                        if current_candle_index == 0 or len(ohlcv_data) > 1:
+                            # push current candle(s)
+                            await self.push(time_frame,
+                                            pair,
+                                            [ohlcv[-1] for ohlcv in ohlcv_data[current_candle_index:]],
+                                            partial=True)
+                    elif not self.pushed_last_init_candles_by_pair[pair]:
+                        # triggered on first iteration to initialize large candles that might be pushed much later
```

Comment:
If it's only triggered during the first iteration, it's better to move this code into an `except` than using a `if` I think.